js/ui/tree_view/ui.tree_view.base.js

```diff
@@ -1158,36 +1158,46 @@ const TreeViewBase = HierarchicalCollectionWidget.inherit({
             return true;
         }
 
-        if(!value && this._isLastRequired(node)) {
+        if(value === false && this._isLastRequired(node)) {
             if(this._showCheckboxes()) {
                 const $node = this._getNodeElement(node);
-                const checkbox = this._getCheckBoxInstance($node);
-
-                checkbox && checkbox.option('value', true);
+                this._setCheckboxValue($node, true);
             }
             return false;
         }
 
-        const selectedNodesKeys = this.getSelectedNodeKeys();
-        if(this._isSingleSelection() && value) {
-            each(selectedNodesKeys, (index, nodeKey) => {
-                this.unselectItem(nodeKey);
+        if(value === true && this._isSingleSelection()) {
+            const selectedKeys = this.getSelectedNodeKeys();
+            each(selectedKeys, (index, key) => {
+                this._dataAdapter.toggleSelection(key, false);
+                this._updateItemsUI();
+
+                this._fireItemSelectionChanged(this._getNode(key));
+                this._fireSelectionChanged();
             });
         }
 
         this._dataAdapter.toggleSelection(node.internalFields.key, value);
         this._updateItemsUI();
 
+        this._fireItemSelectionChanged(node, dxEvent);
+        this._fireSelectionChanged();
+        return true;
+    },
+
+    _setCheckboxValue: function($node, value) {
+        const checkbox = this._getCheckBoxInstance($node);
+        checkbox && checkbox.option('value', value);
+    },
```

Comment:
Will we use this function for "this._getCheckBoxInstance($parentNode).option('value', parentValue);" ?

https://github.com/DevExpress/DevExtreme/blob/21_1/js/ui/tree_view/ui.tree_view.base.js#L1227

Reply:
Yep, I removed the whole method and rewrote code without any checks as in this sample